helix-vcs/src/git.rs

```diff
@@ -0,0 +1,72 @@
+use std::path::Path;
+
+use git_repository::objs::tree::EntryMode;
+use git_repository::{discover, Commit, ObjectId, Repository};
+
+use crate::DiffProvider;
+
+#[cfg(test)]
+mod test;
+
+pub struct Git;
+
+impl DiffProvider for Git {
+    fn get_file_head(&self, file: &Path) -> Option<Vec<u8>> {
+        debug_assert!(file.is_file());
+        debug_assert!(file.is_absolute());
+
+        // discover a repository, requires a directory so we call parent (should not fail but exit gracefully in that case)
+        let repo = discover(file.parent()?).ok()?;
+        let head = repo.head_commit().ok()?;
+        let file_oid = find_file_in_commit(&repo, &head, file)?;
+
+        let file_object = repo.find_object(file_oid).ok()?;
+        Some(file_object.detach().data)
+    }
+}
+
+/// Finds the object that contains the contents of a file at a specific commit.
+fn find_file_in_commit(repo: &Repository, commit: &Commit, file: &Path) -> Option<ObjectId> {
+    let repo_dir = repo.work_dir()?;
+    let rel_path = file.strip_prefix(repo_dir).ok()?;
+    let rel_path_components = byte_path_components(rel_path)?;
+    let tree = commit.tree().ok()?;
+    let tree_entry = tree.lookup_path(rel_path_components).ok()??;
+    match tree_entry.mode {
+        // not a file, everything is new, do not show diff
+        EntryMode::Tree | EntryMode::Commit | EntryMode::Link => None,
+        // found a file
+        EntryMode::Blob | EntryMode::BlobExecutable => Some(tree_entry.oid),
+    }
+}
+
+/// On unix paths are always raw bytes (that are usually utf-8) that can be passed to git directly.
+/// This function is infalliable
+#[cfg(any(unix, target_os = "wasi"))]
+fn byte_path_components(path: &Path) -> Option<impl Iterator<Item = &[u8]>> {
```

Comment:
This is something I want to make easier. Typically here I use  [utilities of the `git-path` crate](https://github.com/Byron/gitoxide/blob/6c064067d2d2cde1c987928c0a44526841f1749f/git-repository/src/revision/spec/parse/delegate/navigate.rs#L127-L132).

I think a utility function in `git-path` could help out with with getting an iterator over components as bytes easier and I will be adding this in v0.24.